## **Task 4.8**

#### **Step 1:** Importing libraries

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Set display options for better viewing

pd.set_option('display.width', 1000)
pd.set_option('display.max_columns', None)  # Limit columns
pd.set_option('display.max_rows', 50)      # Limit rows

In [3]:
# Creating shortcut for data file
path = r'/Users/anjanpakhrin/Documents/Instacart Basket Analysis/'

In [4]:
# Create path to data file --> DataFrame was named after chapter number: "ords_prods_merge_4-7.pkl"

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_4-7.pkl'))

In [5]:
# Checking output

ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  days_since_prior_order  first_order  product_id  add_to_cart_order  reordered                             product_name  aisle_id  department_id  prices    price_range_loc     busiest_day    busiest_days busiest_period_of_day
0   2539329        1             1                  2                  8                     NaN         True         196                  1          0                                     Soda        77              7     9.0  Mid-range product  Regularly busy  Regularly busy        Average orders
1   2539329        1             1                  2                  8                     NaN         True       14084                  2          0  Organic Unsweetened Vanilla Almond Milk        91             16    12.5  Mid-range product  Regularly busy  Regularly busy        Average orders
2   2539329        1             1                  2                  8                     NaN         True       12427                  3          0                      Original Beef Jerky        23             19     4.4  Low-range product  Regularly busy  Regularly busy        Average orders
3   2539329        1             1                  2                  8                     NaN         True       26088                  4          0               Aged White Cheddar Popcorn        23             19     4.7  Low-range product  Regularly busy  Regularly busy        Average orders
4   2539329        1             1                  2                  8                     NaN         True       26405                  5          0         XL Pick-A-Size Paper Towel Rolls        54             17     1.0  Low-range product  Regularly busy  Regularly busy        Average orders

#### **Step 2:** Find aggregated mean of *"order_number"* and grouped by *"department_id"* for **"ords_prods_merge"**

#### Step 2.1: Creating Subset of first one million rows of "ords_prods_merge"

In [6]:
# Create subset "df" of first one million rows

df = ords_prods_merge[:1000000]

In [7]:
# Cheking output for df
df.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  days_since_prior_order  first_order  product_id  add_to_cart_order  reordered                             product_name  aisle_id  department_id  prices    price_range_loc     busiest_day    busiest_days busiest_period_of_day
0   2539329        1             1                  2                  8                     NaN         True         196                  1          0                                     Soda        77              7     9.0  Mid-range product  Regularly busy  Regularly busy        Average orders
1   2539329        1             1                  2                  8                     NaN         True       14084                  2          0  Organic Unsweetened Vanilla Almond Milk        91             16    12.5  Mid-range product  Regularly busy  Regularly busy        Average orders
2   2539329        1             1                  2                  8                     NaN         True       12427                  3          0                      Original Beef Jerky        23             19     4.4  Low-range product  Regularly busy  Regularly busy        Average orders
3   2539329        1             1                  2                  8                     NaN         True       26088                  4          0               Aged White Cheddar Popcorn        23             19     4.7  Low-range product  Regularly busy  Regularly busy        Average orders
4   2539329        1             1                  2                  8                     NaN         True       26405                  5          0         XL Pick-A-Size Paper Towel Rolls        54             17     1.0  Low-range product  Regularly busy  Regularly busy        Average orders

#### Step 2.2: Calculating **means**

In [8]:
# For SUBSET
subset_mean = df.groupby('department_id').agg({'order_number': 'mean'}).round(2)
# Rename the column
subset_mean.columns = ['mean_subset']

# For ENTIRE DATAFRAME
full_mean = ords_prods_merge.groupby('department_id').agg({'order_number': 'mean'}).round(2)
# Rename the column
full_mean.columns = ['mean_full_df']

#### Step 2.3: Comparing: **Subset** vs. **Full** DataFrame

In [9]:
# Combine both dataframes
comparison_df = pd.concat([subset_mean, full_mean], axis=1).reset_index()

# Display the results
comparison_df

department_id  mean_subset  mean_full_df
0               1        14.80         15.46
1               2        17.09         17.28
2               3        17.91         17.17
3               4        17.89         17.81
4               5        15.21         15.22
5               6        15.38         16.44
6               7        17.69         17.23
7               8        16.46         15.34
8               9        15.96         15.90
9              10        20.09         20.20
10             11        16.48         16.17
11             12        15.62         15.89
12             13        16.48         16.58
13             14        17.52         16.77
14             15        15.69         16.17
15             16        18.01         17.67
16             17        16.15         15.69
17             18        19.60         19.31
18             19        17.63         17.18
19             20        17.14         16.47
20             21        21.96         22.90

#### **Step 3:** Analyzing result

In [10]:
# Adding an extra column "difference"to compare values
comparison_df['difference'] = comparison_df['mean_full_df'] - comparison_df['mean_subset']

# Check output
comparison_df

department_id  mean_subset  mean_full_df  difference
0               1        14.80         15.46        0.66
1               2        17.09         17.28        0.19
2               3        17.91         17.17       -0.74
3               4        17.89         17.81       -0.08
4               5        15.21         15.22        0.01
5               6        15.38         16.44        1.06
6               7        17.69         17.23       -0.46
7               8        16.46         15.34       -1.12
8               9        15.96         15.90       -0.06
9              10        20.09         20.20        0.11
10             11        16.48         16.17       -0.31
11             12        15.62         15.89        0.27
12             13        16.48         16.58        0.10
13             14        17.52         16.77       -0.75
14             15        15.69         16.17        0.48
15             16        18.01         17.67       -0.34
16             17        16.15         15.69       -0.46
17             18        19.60         19.31       -0.29
18             19        17.63         17.18       -0.45
19             20        17.14         16.47       -0.67
20             21        21.96         22.90        0.94

#### **Key Notes**
* Most differences are **small** (close to 0), for a subset of the data with the corresponding averages from the full dataset across different departments. That means the subset represents a resonably accurate picture of customer behavior.
* However, several departments, e.g., 6 (1.06), 8 (-1.12), 21 (0.94) show significant discrepancies.
* Some departments, e.g., 5, 9, 13 show **nearly identical means**.
* Overall, the subset tracs closely with the full dataset.

#### **Step 4:** Create **loyalty flag** using *transform()* and *loc()*

#### Creating a flag for customers under following **criteria**:
- If **max_orders > 40**, then **"Loyal customer"**
- If **10 < max_orders <= 40**, then **"Regular customer"**
- If **max_orders <= 10**, then **"New customer"**

1. Split the data into groups based on "user_id" column.
2. Apply **transform()** function on the "order_number" column to generate the maximum orders for each user.
3. Create a new column, "max_order," to put the results of aggregation

#### Step 4.1: Deriving new column with **Transform()**

In [11]:
# Create new column "max_order"
# Group by "user_id"
# Transform "order_number" by applying "max()" function from NumPy library
ords_prods_merge['max_order'] = ords_prods_merge.groupby('user_id')['order_number'].transform('max')

In [12]:
# Checking output - displaying only 2 rows of each user limited to 10 rows in total.

ords_prods_merge.groupby('user_id').head(2).head(10)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  days_since_prior_order  first_order  product_id  add_to_cart_order  reordered                                     product_name  aisle_id  department_id  prices    price_range_loc     busiest_day    busiest_days busiest_period_of_day  max_order
0     2539329        1             1                  2                  8                     NaN         True         196                  1          0                                             Soda        77              7     9.0  Mid-range product  Regularly busy  Regularly busy        Average orders         10
1     2539329        1             1                  2                  8                     NaN         True       14084                  2          0          Organic Unsweetened Vanilla Almond Milk        91             16    12.5  Mid-range product  Regularly busy  Regularly busy        Average orders         10
59    2168274        2             1                  2                 11                     NaN         True       32792                  1          0                   Chipotle Beef & Pork Realstick        23             19     5.2  Mid-range product  Regularly busy  Regularly busy           Most orders         14
60    2168274        2             1                  2                 11                     NaN         True       47766                  2          0                                  Organic Avocado        24              4     6.3  Mid-range product  Regularly busy  Regularly busy           Most orders         14
254   1374495        3             1                  1                 14                     NaN         True        9387                  1          0                              Granny Smith Apples        24              4     8.8  Mid-range product  Regularly busy    Busiest days           Most orders         12
255   1374495        3             1                  1                 14                     NaN         True       17668                  2          0  Unsweetened Chocolate Almond Breeze Almond Milk        91             16     6.6  Mid-range product  Regularly busy    Busiest days           Most orders         12
342   3343014        4             1                  6                 11                     NaN         True       36606                  1          0                       Sprouted Multi-Grain Bread       112              3     9.4  Mid-range product  Regularly busy  Regularly busy           Most orders          5
343   3343014        4             1                  6                 11                     NaN         True        7350                  2          0      Natural Lime Flavor Sparkling Mineral Water       115              7     7.0  Mid-range product  Regularly busy  Regularly busy           Most orders          5
360   2717275        5             1                  3                 12                     NaN         True       15349                  1          0                         Organic Raw Agave Nectar        29             13    10.7  Mid-range product  Regularly busy    Slowest days           Most orders          4
361   2717275        5             1                  3                 12                     NaN         True       21413                  2          0                                     Organic Soba        66              6     6.7  Mid-range product  Regularly busy    Slowest days           Most orders          4

#### Step 4.2: Deriving Columns with **loc()**
#### Creating new flagging column **loyalty_flag**

In [13]:
# If max_order > 40, then "Loyal customer"
ords_prods_merge.loc[ords_prods_merge['max_order'] > 15, 'loyalty_flag'] = 'Loyal customer'

# If 10 < max_orders <= 40, then "Regular customer"
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order']> 10), 'loyalty_flag'] = 'Regular customer'

# If max_orders <= 10, then "New customer"
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 10, 'loyalty_flag')] = 'New customer'

In [14]:
# Checking output for new column - displaying only 2 rows per category
ords_prods_merge.groupby('loyalty_flag').head(2)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  days_since_prior_order  first_order  product_id  add_to_cart_order  reordered                             product_name  aisle_id  department_id  prices    price_range_loc     busiest_day    busiest_days busiest_period_of_day  max_order      loyalty_flag
0      2539329        1             1                  2                  8                     NaN         True         196                  1          0                                     Soda        77              7     9.0  Mid-range product  Regularly busy  Regularly busy        Average orders         10      New customer
1      2539329        1             1                  2                  8                     NaN         True       14084                  2          0  Organic Unsweetened Vanilla Almond Milk        91             16    12.5  Mid-range product  Regularly busy  Regularly busy        Average orders         10      New customer
59     2168274        2             1                  2                 11                     NaN         True       32792                  1          0           Chipotle Beef & Pork Realstick        23             19     5.2  Mid-range product  Regularly busy  Regularly busy           Most orders         14  Regular customer
60     2168274        2             1                  2                 11                     NaN         True       47766                  2          0                          Organic Avocado        24              4     6.3  Mid-range product  Regularly busy  Regularly busy           Most orders         14  Regular customer
2474    361493       27             1                  3                  9                     NaN         True       25718                  1          0                         Cold Brew Coffee        26              7    10.4  Mid-range product  Regularly busy    Slowest days           Most orders         81    Loyal customer
2475   1662354       27             2                  2                 17                     6.0        False       30776                  1          0           Organic Raw Kombucha Gingerade        31              7    13.4  Mid-range product  Regularly busy  Regularly busy        Average orders         81    Loyal customer

In [15]:
# Checking frequency for each category
ords_prods_merge['loyalty_flag'].value_counts(dropna=False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

**Majority of orders are placed by "Regular customer". Second highest orders are done by "Loyal customer". The lowest numbers of orders are placed by "New customer".** 

In [16]:
# Counting the number of customers grouped by "loyalty_flag"

ords_prods_merge.groupby(['loyalty_flag']).user_id.nunique()

loyalty_flag
Loyal customer       17017
New customer        112328
Regular customer     76864
Name: user_id, dtype: int64

**"New customers" constitute Instacart's largest segment, followed by "Loyal customers" as the second-largest group, and "Regular customers" as the third-largest.**

#### **Step 5:** Determining purchasing price difference by customer category - **loyalty_flag**

In [17]:
# Step 1: Calculating key summary statistics grouped by "loyalty_flag"
loyalty_price_stats = ords_prods_merge.groupby('loyalty_flag')['prices'].agg(['mean', 'median', 'min', 'max', 'std'])

# Return statistics
loyalty_price_stats

mean  median  min      max         std
loyalty_flag                                                 
Loyal customer    10.386335     7.4  1.0  99999.0  328.017792
New customer      13.294669     7.4  1.0  99999.0  597.560303
Regular customer  12.495717     7.4  1.0  99999.0  539.720947

#### Analysis of purchasing behavior based on prices for different customer categories
* Based on above statistics, **loyal customers spend less** compared to new and regular customer.
* New customers spend the most.
* Same median values (7.4) suggest that the majority of purchases is very similar, regardless of loyalty status.
* Significant difference in **min** and **max** for all groups shows that the dataset contains some **extrem outliers**, which might be data entry errors or rare products prices. This obviously affects the mean and standard deviation.
* Standard deviation is much higher for new and regular customers, This means New and regular customers' spending behavior is less predictable. They might be trying a wide variety of products at very different price points. In opposite, loyal customers have likely settled into a routine of buying thier regular, known products.

#### **Step 6:** Identifying different type of spenders under following criteria:
* **Low spender:** Users with mean product price < 10.
  * **Condition:** user_mean_prices < 10
* **High spender:** Users wuth mean product price is greater or equal to 10.
  * **Condition:** user_mean_prices >= 10

#### Step 6.1: Aggregating with transform() --> Creating **"user mean prices"**
* Split data into groups based on **"user_id"**.
* Apply **transform()** on **"prices"** column to generate the mean of the prices of products purchased.
* Create a new column **"user_mean_prices"** to put the results

In [18]:
# Create ne column "user_mean_prices"
# Group by "user_id"
# Transform "prices" by "mean()" function

ords_prods_merge['user_mean_prices'] = ords_prods_merge.groupby('user_id')['prices'].transform('mean')

In [19]:
# Checking output

ords_prods_merge.groupby('user_id').head(2).head(10)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  days_since_prior_order  first_order  product_id  add_to_cart_order  reordered                                     product_name  aisle_id  department_id  prices    price_range_loc     busiest_day    busiest_days busiest_period_of_day  max_order      loyalty_flag  user_mean_prices
0     2539329        1             1                  2                  8                     NaN         True         196                  1          0                                             Soda        77              7     9.0  Mid-range product  Regularly busy  Regularly busy        Average orders         10      New customer          6.367797
1     2539329        1             1                  2                  8                     NaN         True       14084                  2          0          Organic Unsweetened Vanilla Almond Milk        91             16    12.5  Mid-range product  Regularly busy  Regularly busy        Average orders         10      New customer          6.367797
59    2168274        2             1                  2                 11                     NaN         True       32792                  1          0                   Chipotle Beef & Pork Realstick        23             19     5.2  Mid-range product  Regularly busy  Regularly busy           Most orders         14  Regular customer          7.515897
60    2168274        2             1                  2                 11                     NaN         True       47766                  2          0                                  Organic Avocado        24              4     6.3  Mid-range product  Regularly busy  Regularly busy           Most orders         14  Regular customer          7.515897
254   1374495        3             1                  1                 14                     NaN         True        9387                  1          0                              Granny Smith Apples        24              4     8.8  Mid-range product  Regularly busy    Busiest days           Most orders         12  Regular customer          8.197727
255   1374495        3             1                  1                 14                     NaN         True       17668                  2          0  Unsweetened Chocolate Almond Breeze Almond Milk        91             16     6.6  Mid-range product  Regularly busy    Busiest days           Most orders         12  Regular customer          8.197727
342   3343014        4             1                  6                 11                     NaN         True       36606                  1          0                       Sprouted Multi-Grain Bread       112              3     9.4  Mid-range product  Regularly busy  Regularly busy           Most orders          5      New customer          8.205555
343   3343014        4             1                  6                 11                     NaN         True        7350                  2          0      Natural Lime Flavor Sparkling Mineral Water       115              7     7.0  Mid-range product  Regularly busy  Regularly busy           Most orders          5      New customer          8.205555
360   2717275        5             1                  3                 12                     NaN         True       15349                  1          0                         Organic Raw Agave Nectar        29             13    10.7  Mid-range product  Regularly busy    Slowest days           Most orders          4      New customer          9.189189
361   2717275        5             1                  3                 12                     NaN         True       21413                  2          0                                     Organic Soba        66              6     6.7  Mid-range product  Regularly busy    Slowest days           Most orders          4      New customer          9.189189

#### **Summary to above table**
* Appended a new column **"user_mean_prices"** based on mean user item price.
* Calculated **mean price per user** and assigned the value to all user rows.
* Diplayed 2 rows per user for the first 5 users, limited to a total 10 rows for a **preliminary review** of the new column.

#### Step 6.2: Deriving column with loc() --> Creating **"Spending flag"**
#### Criteria:
* If **user_mean_prices** < 10, then **"Low spender"**
* if **user_mean_prices** >= 10, then **"High spender"**

In [20]:
# If user_mean_prices < 10, then "Low spender"
ords_prods_merge.loc[ords_prods_merge['user_mean_prices'] < 10, 'spending_flag'] = 'Low spender'

# If user_mean_prices >= 10, then "High spender"
ords_prods_merge.loc[ords_prods_merge['user_mean_prices'] >= 10, 'spending_flag'] = 'High spender'

In [21]:
# Checking output for new column
ords_prods_merge.groupby('spending_flag').head(5)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  days_since_prior_order  first_order  product_id  add_to_cart_order  reordered                                 product_name  aisle_id  department_id  prices    price_range_loc     busiest_day    busiest_days busiest_period_of_day  max_order      loyalty_flag  user_mean_prices spending_flag
0      2539329        1             1                  2                  8                     NaN         True         196                  1          0                                         Soda        77              7     9.0  Mid-range product  Regularly busy  Regularly busy        Average orders         10      New customer          6.367797   Low spender
1      2539329        1             1                  2                  8                     NaN         True       14084                  2          0      Organic Unsweetened Vanilla Almond Milk        91             16    12.5  Mid-range product  Regularly busy  Regularly busy        Average orders         10      New customer          6.367797   Low spender
2      2539329        1             1                  2                  8                     NaN         True       12427                  3          0                          Original Beef Jerky        23             19     4.4  Low-range product  Regularly busy  Regularly busy        Average orders         10      New customer          6.367797   Low spender
3      2539329        1             1                  2                  8                     NaN         True       26088                  4          0                   Aged White Cheddar Popcorn        23             19     4.7  Low-range product  Regularly busy  Regularly busy        Average orders         10      New customer          6.367797   Low spender
4      2539329        1             1                  2                  8                     NaN         True       26405                  5          0             XL Pick-A-Size Paper Towel Rolls        54             17     1.0  Low-range product  Regularly busy  Regularly busy        Average orders         10      New customer          6.367797   Low spender
1486   1737705       17             1                  2                 13                     NaN         True        7350                  1          0  Natural Lime Flavor Sparkling Mineral Water       115              7     7.0  Mid-range product  Regularly busy  Regularly busy           Most orders         40  Regular customer        108.648300  High spender
1487   1737705       17             1                  2                 13                     NaN         True       47141                  2          0                                         Cola        77              7    14.6  Mid-range product  Regularly busy  Regularly busy           Most orders         40  Regular customer        108.648300  High spender
1488   1737705       17             1                  2                 13                     NaN         True       17762                  3          0                       Light Oaked Chardonnay        62              5     7.0  Mid-range product  Regularly busy  Regularly busy           Most orders         40  Regular customer        108.648300  High spender
1489   1681401       17             2                  5                 10                     3.0        False        7350                  1          1  Natural Lime Flavor Sparkling Mineral Water       115              7     7.0  Mid-range product  Regularly busy  Regularly busy           Most orders         40  Regular customer        108.648300  High spender
1490   1681401       17             2                  5                 10                     3.0        False       18534                  2          0                     Grade A Extra Large Eggs        86             16     8.0  Mid-range product  Regularly busy  Regularly busy           Most orders         40  Regular customer  

In [22]:
# Checking frequncy of each category
ords_prods_merge['spending_flag'].value_counts(dropna=False)

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

#### **Key Observations:**
* **Extreme Imbalance:** **98%** of all orders fall under **"Low spender"** category (31769965/(31769965+634894) = 0.98). This is not a normal or expected distribution average item price.
* **Implausible Average:** A **108.65** average item cost for **"High spender"** users is unreasllistically high for grocery products, indication a data anomaly, not genuine consumer behavior.
* **Outlier Confirmation:** The drastic gap between the **"Low spender" (6.37)** and **"High spender" (108.65)** means confirms that extreme outliers are artificially infalting the averages for small user user group.

#### **Recommendations:**
* Investigate the **prices** column and conduct the data cleaning.
* **Re-running** of analysis suggested.  

In [23]:
# Counting the number of customers grouped by "spending_flag"

ords_prods_merge.groupby(['spending_flag']).user_id.nunique()

spending_flag
High spender      5357
Low spender     200852
Name: user_id, dtype: int64

#### **Summary:**
* **Low spenders** dominate the customer base, comprising **200,852** customers compared to only **5,357 high spenders.**
* **High spenders** represent a niche segment, making up just **2.6%** of the total customer population while **low spenders** constitute the overwhelming **majority at 97.4%**

In [24]:
# Checking few statistics
pd.set_option('display.float_format', lambda x: '%.2f' % x)
ords_prods_merge['prices'].describe()

count   32404859.00
mean          11.98
std          495.64
min            1.00
25%            4.20
50%            7.40
75%           11.30
max        99999.00
Name: prices, dtype: float64

#### **Possible Outlier**: 99999.00

#### **Addressing the outlier prices**

In [25]:
# Identifying outliers
outlier_count = len(ords_prods_merge[ords_prods_merge['prices'] > 100])
total_count = len(ords_prods_merge)

print(f"Number of records with price > $100: {outlier_count}")
print(f"Total number of records: {total_count}")

Number of records with price > $100: 5127
Total number of records: 32404859


In [26]:
# Find all unique prices greater than 100 and sort them
outlier_prices = ords_prods_merge.loc[ords_prods_merge['prices'] > 100, 'prices'].unique()
outlier_prices_sorted = np.sort(outlier_prices)[::-1] # Sort in descending order

print("Unique outlier prices (>$100):")
print(outlier_prices_sorted)

Unique outlier prices (>$100):
[99999. 14900.]


#### **Replacing outlier products**

In [27]:
ords_prods_merge['prices'] = ords_prods_merge['prices'].replace({
    99999: 9.2,
    14900: 8
})

In [28]:
ords_prods_merge['prices'].describe()

count   32404859.00
mean           7.79
std            4.10
min            1.00
25%            4.20
50%            7.40
75%           11.30
max           25.00
Name: prices, dtype: float64

## **Re-anylysis for Step 5 - 6**

#### **Re-Step 5:** Determining purchasing price difference by customer category - **loyalty_flag**

In [29]:
# Step 1: Calculating key summary statistics grouped by "loyalty_flag"
loyalty_price_stats = ords_prods_merge.groupby('loyalty_flag')['prices'].agg(['mean', 'median', 'min', 'max', 'std'])

# Return statistics
loyalty_price_stats

mean  median  min   max  std
loyalty_flag                                  
Loyal customer    7.77    7.40 1.00 25.00 4.21
New customer      7.80    7.40 1.00 25.00 4.27
Regular customer  7.80    7.40 1.00 25.00 4.25

#### Analysis of purchasing behavior based on prices for different customer categories
* Based on above statistics, **loyal customers spend less** compared to new and regular customer.
* Same median values (7.4) suggest that the majority of purchases is very similar, regardless of loyalty status.

#### **Dropping columns: "spending_flag", "median_frequency", "frequency_flag" for re-analysis**

In [30]:
# Checking dropping columns
ords_prods_merge.head(1)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  days_since_prior_order  first_order  product_id  add_to_cart_order  reordered product_name  aisle_id  department_id  prices    price_range_loc     busiest_day    busiest_days busiest_period_of_day  max_order  loyalty_flag  user_mean_prices spending_flag
0   2539329        1             1                  2                  8                     NaN         True         196                  1          0         Soda        77              7    9.00  Mid-range product  Regularly busy  Regularly busy        Average orders         10  New customer              6.37   Low spender

#### **Re-Step 6:** Identifying different type of spenders under following criteria:
* **Low spender:** Users with mean product price < 10.
  * **Condition:** user_mean_prices < 10
* **High spender:** Users wuth mean product price is greater or equal to 10.
  * **Condition:** user_mean_prices >= 10
#### Re-Step 6.1: Aggregating with transform() --> Creating **"user mean prices"**
* Split data into groups based on **"user_id"**.
* Apply **transform()** on **"prices"** column to generate the mean of the prices of products purchased.
* Create a new column **"user_mean_prices"** to put the results

In [31]:
# Create ne column "user_mean_prices"
# Group by "user_id"
# Transform "prices" by "mean()" function

ords_prods_merge['user_mean_prices'] = ords_prods_merge.groupby('user_id')['prices'].transform('mean')

In [32]:
# Checking output

ords_prods_merge.groupby('user_id').head(2).head(10)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  days_since_prior_order  first_order  product_id  add_to_cart_order  reordered                                     product_name  aisle_id  department_id  prices    price_range_loc     busiest_day    busiest_days busiest_period_of_day  max_order      loyalty_flag  user_mean_prices spending_flag
0     2539329        1             1                  2                  8                     NaN         True         196                  1          0                                             Soda        77              7    9.00  Mid-range product  Regularly busy  Regularly busy        Average orders         10      New customer              6.37   Low spender
1     2539329        1             1                  2                  8                     NaN         True       14084                  2          0          Organic Unsweetened Vanilla Almond Milk        91             16   12.50  Mid-range product  Regularly busy  Regularly busy        Average orders         10      New customer              6.37   Low spender
59    2168274        2             1                  2                 11                     NaN         True       32792                  1          0                   Chipotle Beef & Pork Realstick        23             19    5.20  Mid-range product  Regularly busy  Regularly busy           Most orders         14  Regular customer              7.52   Low spender
60    2168274        2             1                  2                 11                     NaN         True       47766                  2          0                                  Organic Avocado        24              4    6.30  Mid-range product  Regularly busy  Regularly busy           Most orders         14  Regular customer              7.52   Low spender
254   1374495        3             1                  1                 14                     NaN         True        9387                  1          0                              Granny Smith Apples        24              4    8.80  Mid-range product  Regularly busy    Busiest days           Most orders         12  Regular customer              8.20   Low spender
255   1374495        3             1                  1                 14                     NaN         True       17668                  2          0  Unsweetened Chocolate Almond Breeze Almond Milk        91             16    6.60  Mid-range product  Regularly busy    Busiest days           Most orders         12  Regular customer              8.20   Low spender
342   3343014        4             1                  6                 11                     NaN         True       36606                  1          0                       Sprouted Multi-Grain Bread       112              3    9.40  Mid-range product  Regularly busy  Regularly busy           Most orders          5      New customer              8.21   Low spender
343   3343014        4             1                  6                 11                     NaN         True        7350                  2          0      Natural Lime Flavor Sparkling Mineral Water       115              7    7.00  Mid-range product  Regularly busy  Regularly busy           Most orders          5      New customer              8.21   Low spender
360   2717275        5             1                  3                 12                     NaN         True       15349                  1          0                         Organic Raw Agave Nectar        29             13   10.70  Mid-range product  Regularly busy    Slowest days           Most orders          4      New customer              9.19   Low spender
361   2717275        5             1                  3                 12                     NaN         True       21413                  2          0                                     Organic Soba        66              6    6.70  Mid-range product  Regularly busy    Slowest days           Most ord

#### **Summary to above table**
* Appended a new column **"user_mean_prices"** based on mean user item price.
* Calculated **mean price per user** and assigned the value to all user rows.
* Diplayed 2 rows per user for the first 5 users, limited to a total 10 rows for a **preliminary review** of the new column.

#### Re-Step 6.2: Deriving column with loc() --> Creating **"Spending flag"**
#### Criteria:
* If **user_mean_prices** < 10, then **"Low spender"**
* if **user_mean_prices** >= 10, then **"High spender"**

In [33]:
# If user_mean_prices < 10, then "Low spender"
ords_prods_merge.loc[ords_prods_merge['user_mean_prices'] < 10, 'spending_flag'] = 'Low spender'

# If user_mean_prices >= 10, then "High spender"
ords_prods_merge.loc[ords_prods_merge['user_mean_prices'] >= 10, 'spending_flag'] = 'High spender'

In [34]:
# Checking output for new column
ords_prods_merge.groupby('spending_flag').head(5)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  days_since_prior_order  first_order  product_id  add_to_cart_order  reordered                             product_name  aisle_id  department_id  prices    price_range_loc     busiest_day    busiest_days busiest_period_of_day  max_order      loyalty_flag  user_mean_prices spending_flag
0      2539329        1             1                  2                  8                     NaN         True         196                  1          0                                     Soda        77              7    9.00  Mid-range product  Regularly busy  Regularly busy        Average orders         10      New customer              6.37   Low spender
1      2539329        1             1                  2                  8                     NaN         True       14084                  2          0  Organic Unsweetened Vanilla Almond Milk        91             16   12.50  Mid-range product  Regularly busy  Regularly busy        Average orders         10      New customer              6.37   Low spender
2      2539329        1             1                  2                  8                     NaN         True       12427                  3          0                      Original Beef Jerky        23             19    4.40  Low-range product  Regularly busy  Regularly busy        Average orders         10      New customer              6.37   Low spender
3      2539329        1             1                  2                  8                     NaN         True       26088                  4          0               Aged White Cheddar Popcorn        23             19    4.70  Low-range product  Regularly busy  Regularly busy        Average orders         10      New customer              6.37   Low spender
4      2539329        1             1                  2                  8                     NaN         True       26405                  5          0         XL Pick-A-Size Paper Towel Rolls        54             17    1.00  Low-range product  Regularly busy  Regularly busy        Average orders         10      New customer              6.37   Low spender
9425   2451656       72             1                  5                 15                     NaN         True       38689                  1          0                 Organic Reduced Fat Milk        84             16    8.80  Mid-range product  Regularly busy  Regularly busy           Most orders         12  Regular customer             10.03  High spender
9426   2451656       72             1                  5                 15                     NaN         True       32553                  2          0         Country Stand Juice, Medium Pulp        98              7   14.80  Mid-range product  Regularly busy  Regularly busy           Most orders         12  Regular customer             10.03  High spender
9427   2451656       72             1                  5                 15                     NaN         True       46906                  3          0               Grape White/Green Seedless        24              4    7.20  Mid-range product  Regularly busy  Regularly busy           Most orders         12  Regular customer             10.03  High spender
9428   2451656       72             1                  5                 15                     NaN         True       27992                  4          0    Unbleached Recycled Paper Towel Rolls        54             17    9.20  Mid-range product  Regularly busy  Regularly busy           Most orders         12  Regular customer             10.03  High spender
9429   2451656       72             1                  5                 15                     NaN         True       26369                  5          0                      Organic Roma Tomato        83              4   15.00  Mid-range product  Regularly busy  Regularly busy           Most orders         12  Regular customer             10.03  High spender

In [35]:
# Checking frequncy of each category
ords_prods_merge['spending_flag'].value_counts(dropna=False)

spending_flag
Low spender     32285150
High spender      119709
Name: count, dtype: int64

#### **Number of customers based onf spending**

In [36]:
# Counting the number of customers grouped by "spending_flag"

ords_prods_merge.groupby(['spending_flag']).user_id.nunique()

spending_flag
High spender      3388
Low spender     202821
Name: user_id, dtype: int64

#### **Summary:**
* **Vast majority are low-spenders**, indicating a business model reliant on high-volume, small transactions.
* **Small high-spender** segment is critically important, likely generating a disproportionate share of total revenue.

#### **Step 7:** Identifying **frequent**, **non-frequent**, and **regular** customers
#### Criteria:
* If median of **days_since_prior_order > 20**, then **"Non-frequent customer"**
* If median of **10 < days_since_prior_order <= 20**, then **"Regular customer"**
* If median of **days_since_prior_order <= 10**, then **"Frequent customer"**

#### Step 7.1: Calculation **median** for **days_since_prior_order**

In [37]:
# Create new column "median_frequency" for median for days_since_prior_order
# Group by "user_id"
# Transform "days_since_prior_order" by applying "max()" function
ords_prods_merge['median_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform('median')

In [38]:
# Checking output

ords_prods_merge.groupby('user_id').head(2).head(10)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  days_since_prior_order  first_order  product_id  add_to_cart_order  reordered                                     product_name  aisle_id  department_id  prices    price_range_loc     busiest_day    busiest_days busiest_period_of_day  max_order      loyalty_flag  user_mean_prices spending_flag  median_frequency
0     2539329        1             1                  2                  8                     NaN         True         196                  1          0                                             Soda        77              7    9.00  Mid-range product  Regularly busy  Regularly busy        Average orders         10      New customer              6.37   Low spender             20.50
1     2539329        1             1                  2                  8                     NaN         True       14084                  2          0          Organic Unsweetened Vanilla Almond Milk        91             16   12.50  Mid-range product  Regularly busy  Regularly busy        Average orders         10      New customer              6.37   Low spender             20.50
59    2168274        2             1                  2                 11                     NaN         True       32792                  1          0                   Chipotle Beef & Pork Realstick        23             19    5.20  Mid-range product  Regularly busy  Regularly busy           Most orders         14  Regular customer              7.52   Low spender             13.00
60    2168274        2             1                  2                 11                     NaN         True       47766                  2          0                                  Organic Avocado        24              4    6.30  Mid-range product  Regularly busy  Regularly busy           Most orders         14  Regular customer              7.52   Low spender             13.00
254   1374495        3             1                  1                 14                     NaN         True        9387                  1          0                              Granny Smith Apples        24              4    8.80  Mid-range product  Regularly busy    Busiest days           Most orders         12  Regular customer              8.20   Low spender             10.00
255   1374495        3             1                  1                 14                     NaN         True       17668                  2          0  Unsweetened Chocolate Almond Breeze Almond Milk        91             16    6.60  Mid-range product  Regularly busy    Busiest days           Most orders         12  Regular customer              8.20   Low spender             10.00
342   3343014        4             1                  6                 11                     NaN         True       36606                  1          0                       Sprouted Multi-Grain Bread       112              3    9.40  Mid-range product  Regularly busy  Regularly busy           Most orders          5      New customer              8.21   Low spender             20.00
343   3343014        4             1                  6                 11                     NaN         True        7350                  2          0      Natural Lime Flavor Sparkling Mineral Water       115              7    7.00  Mid-range product  Regularly busy  Regularly busy           Most orders          5      New customer              8.21   Low spender             20.00
360   2717275        5             1                  3                 12                     NaN         True       15349                  1          0                         Organic Raw Agave Nectar        29             13   10.70  Mid-range product  Regularly busy    Slowest days           Most orders          4      New customer              9.19   Low spender             11.00
361   2717275        5             1                  3                 12                     NaN         True       21413  

#### Step 7.2: Creating flagging column **frequency_flag**

In [39]:
# If median_frequency > 20, then "Non-frequent customer"
ords_prods_merge.loc[ords_prods_merge['median_frequency'] > 20, 'frequency_flag'] = 'Non-frequent customer'

# If 10 < median_days_since_prior_order <= 20, then "Regular customer"
ords_prods_merge.loc[(ords_prods_merge['median_frequency'] <= 20) & (ords_prods_merge['median_frequency']> 10), 'frequency_flag'] = 'Regular customer'

# If max_orders <= 10, then "New customer"
ords_prods_merge.loc[(ords_prods_merge['median_frequency'] <= 10, 'frequency_flag')] = 'Frequent customer'

In [40]:
# Checking output for new column
ords_prods_merge.groupby('frequency_flag').head(2)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  days_since_prior_order  first_order  product_id  add_to_cart_order  reordered                                     product_name  aisle_id  department_id  prices    price_range_loc     busiest_day    busiest_days busiest_period_of_day  max_order      loyalty_flag  user_mean_prices spending_flag  median_frequency         frequency_flag
0     2539329        1             1                  2                  8                     NaN         True         196                  1          0                                             Soda        77              7    9.00  Mid-range product  Regularly busy  Regularly busy        Average orders         10      New customer              6.37   Low spender             20.50  Non-frequent customer
1     2539329        1             1                  2                  8                     NaN         True       14084                  2          0          Organic Unsweetened Vanilla Almond Milk        91             16   12.50  Mid-range product  Regularly busy  Regularly busy        Average orders         10      New customer              6.37   Low spender             20.50  Non-frequent customer
59    2168274        2             1                  2                 11                     NaN         True       32792                  1          0                   Chipotle Beef & Pork Realstick        23             19    5.20  Mid-range product  Regularly busy  Regularly busy           Most orders         14  Regular customer              7.52   Low spender             13.00       Regular customer
60    2168274        2             1                  2                 11                     NaN         True       47766                  2          0                                  Organic Avocado        24              4    6.30  Mid-range product  Regularly busy  Regularly busy           Most orders         14  Regular customer              7.52   Low spender             13.00       Regular customer
254   1374495        3             1                  1                 14                     NaN         True        9387                  1          0                              Granny Smith Apples        24              4    8.80  Mid-range product  Regularly busy    Busiest days           Most orders         12  Regular customer              8.20   Low spender             10.00      Frequent customer
255   1374495        3             1                  1                 14                     NaN         True       17668                  2          0  Unsweetened Chocolate Almond Breeze Almond Milk        91             16    6.60  Mid-range product  Regularly busy    Busiest days           Most orders         12  Regular customer              8.20   Low spender             10.00      Frequent customer

#### Step 7.3: Analyzing results

In [41]:
# Checking frequency for each category
ords_prods_merge['frequency_flag'].value_counts(dropna=False)

frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [42]:
# Counting customers based on "frequency_flag"

ords_prods_merge.groupby(['frequency_flag']).user_id.nunique()

frequency_flag
Frequent customer        86596
Non-frequent customer    59619
Regular customer         59993
Name: user_id, dtype: int64

* **Frequent customers form the largest segment** with **86,596** users, indicating a strong core user base that relies consistently on the platform for shopping needs.
* **Regular customers (59,993)** and **non-frequent customers (59,619)** represent significant segments of similar size, highlighting key opportunities for targeted engagement strategies to increase purchase frequency.
* These **three distinct customer categories** enable precise marketing approaches and personalized notifications designed to encourage more frequent purchasing behavior across all user types.

#### **Step 9:** Exporting dataframe

In [43]:
# Export data as pickle
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_4-8.pkl'))